L1Trigger/L1TGEM/test/GEMTriggerPrimitivesAnalyzer.cc

```diff
@@ -0,0 +1,229 @@
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "CommonTools/UtilAlgos/interface/TFileService.h"
+
+#include "TH1F.h"
+#include "TPostScript.h"
+#include "TCanvas.h"
+#include "TFile.h"
+#include "TText.h"
+#include "TPaveLabel.h"
+#include "TLegend.h"
+#include "TStyle.h"
+#include "TROOT.h"
+
+#include <iostream>
+#include <string>
+
+class GEMTriggerPrimitivesAnalyzer : public edm::one::EDAnalyzer<> {
+public:
+  /// Constructor
+  explicit GEMTriggerPrimitivesAnalyzer(const edm::ParameterSet &conf);
+
+  /// Destructor
+  ~GEMTriggerPrimitivesAnalyzer() override {}
+
+  /// Does the job
+  void analyze(const edm::Event &event, const edm::EventSetup &setup) override;
+
+  /// Write to ROOT file, make plots, etc.
+  void endJob() override;
+
+private:
+  void makePlot(TH1F *dataMon,
+                TH1F *emulMon,
+                TH1F *diffMon,
+                TString lcts,
+                TString lct,
+                TString var,
+                TString chamber,
+                TPostScript *ps,
+                TCanvas *c1) const;
+
+  // plots of data vs emulator
+  std::string rootFileName_;
+  unsigned runNumber_;
+  std::string monitorDir_;
+  std::vector<std::string> chambers_;
+  std::vector<std::string> clusterVars_;
+  bool dataVsEmulatorPlots_;
+  void makeDataVsEmulatorPlots();
+
+  // plots of efficiencies in MC
+  bool mcEfficiencyPlots_;
+
+  // plots of resolution in MC
+  bool mcResolutionPlots_;
+
+  /*
+    When set to True, we assume that the data comes from
+    the Building 904 GEM test-stand. This test-stand is a single
+    ME1/1 chamber.
+  */
+  bool B904Setup_;
+  // label only relevant for B904 local runs
+  std::string B904RunNumber_;
+};
+
+GEMTriggerPrimitivesAnalyzer::GEMTriggerPrimitivesAnalyzer(const edm::ParameterSet &conf)
+    : rootFileName_(conf.getParameter<std::string>("rootFileName")),
+      runNumber_(conf.getParameter<unsigned>("runNumber")),
+      monitorDir_(conf.getParameter<std::string>("monitorDir")),
+      chambers_(conf.getParameter<std::vector<std::string>>("chambers")),
+      clusterVars_(conf.getParameter<std::vector<std::string>>("clusterVars")),
+      dataVsEmulatorPlots_(conf.getParameter<bool>("dataVsEmulatorPlots")),
+      mcEfficiencyPlots_(conf.getParameter<bool>("mcEfficiencyPlots")),
+      mcResolutionPlots_(conf.getParameter<bool>("mcResolutionPlots")),
+      B904Setup_(conf.getParameter<bool>("B904Setup")),
+      B904RunNumber_(conf.getParameter<std::string>("B904RunNumber")) {}
+
+void GEMTriggerPrimitivesAnalyzer::analyze(const edm::Event &ev, const edm::EventSetup &setup) {
+  // efficiency and resolution analysis is done here
+}
+
+void GEMTriggerPrimitivesAnalyzer::endJob() {
+  if (dataVsEmulatorPlots_)
+    makeDataVsEmulatorPlots();
+}
+
+void GEMTriggerPrimitivesAnalyzer::makeDataVsEmulatorPlots() {
+  // data vs emulator plots are created here
+  edm::Service<TFileService> fs;
```

Comment:
Because this uses the TFileService which is not thread safe, you must declare this as using a SharedResource and explicitly set the resource in the constructor.

Reply:
Done.